README.md

```diff
@@ -59,6 +59,58 @@ Example: Instead of `file-server.ts` use `file_server.ts`.
 More specifically, code should be wrapped at 80 columns and use 2-space
 indentation and use camel-case. Use `//format.ts` to invoke prettier.
 
+### Exported functions: max 2 args, put rest into an options object.
+
+When designing APIs, stick to the following rules.
+
+1. A function that is part of the public API takes 0-2 'normal' arguments,
+   plus (if necessary) an options object (so max 3 total).
+
+2. Optional parameters _always_ go into the options object.
+
+3. The 'options' argument is the only argument that is a regular 'Object'.
+
+   Other arguments can be objects, but they must be distinguishable from a
+   'plain' Object runtime, by having either:
+
+   - a distinguishing prototype (e.g. `Array`, `Map`, `Date`, `class MyThing`)
+   - a well-known symbol property (e.g. an iterable with `Symbol.iterator`).
+
+```ts
+// BAD: too many arguments. (#1)
+export function write(
+  fd: number,
+  buffer: TypedArray,
+  offset: number,
+  length: number,
+  position: number
+): void {}
+
+// BAD: optional argument not part of options object. (#2)
+export function resolveIP(
+  hostname: string,
+  family?: "ipv4" | "ipv6"
+): IPAddress[] {}
+
+// BAD: `env` can be a regular Object. (#3)
+export function runShell(
+  cmdline: string,
+  env: { [key: string]: string },
+  options: RunShellOptions = {}
+): string {}
+
+// GOOD.
+interface RenameOptions {
+  replaceExisting?: boolean;
+  followLinks?: boolean;
+}
+export function renameSync(
+  oldname: string,
+  newname: string,
+  options: RenameOptions = {}
+) {}
+```
```

Comment:
I think it's nice if you have the GOOD and BAD examples correspond, so people see how to redesign.